### PR TITLE
Module to manage F5 GTM virtual servers

### PR DESCRIPTION
**Pull request by perzizzle**
Supports enabling/disabling F5 GTM virtual servers. Stubbed out where to add/remove virtual servers.

### CODE REVIEW

**Comment by gregdek** [comment]
@perzizzle ready for review?

**Comment by perzizzle** [comment]
I think its ready for review. There is plenty of opportunity to add more features but this is all I needed so far. I intend on adding more eventually.

**Comment by gregdek** [comment]
@mhite @srvg as the owners of the other F5 modules, can you guys review this new module?

**Comment by mhite** [comment]
Since I don't run GTM, I'm a bad person to test this. Can you solicit some GTM users on the mailing list to test this?

File: network/f5/bigip_gtm_virtual_server.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+'''
+
+try:
+    import bigsuds
+except ImportError:
+    bigsuds_found = False
+else:
+    bigsuds_found = True
+
+def bigip_api(bigip, user, password):
```
Comment (srgvg) [inline network/f5/bigip_gtm_virtual_server.py]: Perhaps replace `bigip` with `server` for consistency

File: network/f5/bigip_gtm_virtual_server.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    result = {'changed': False}  # default
+
+    try:
+        api = bigip_api(server, user, password)
+
+        if not virtual_server_exists(api, virtual_server_name, virtual_server_server):
+            module.fail_json(msg="virtual server does not exist")
+
+        if state == 'absent':
+            if virtual_server_name and virtual_server_server:
```
Comment (srgvg) [inline network/f5/bigip_gtm_virtual_server.py]: This condition seems unneccessary as both vars are required?
Also, _if_ this condition wasn't met (which can't happen), the module would exit without result being set.

File: network/f5/bigip_gtm_virtual_server.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    address = module.params['host']
+    port = module.params['port']
+
+    result = {'changed': False}  # default
+
+    try:
+        api = bigip_api(server, user, password)
+
+        if not virtual_server_exists(api, virtual_server_name, virtual_server_server):
+            module.fail_json(msg="virtual server does not exist")
```
Comment (srgvg) [inline network/f5/bigip_gtm_virtual_server.py]: Why is this here? Module fails if the virtual server does not exist. But later on you do similar checks depending on when it should be created etc.

File: network/f5/bigip_gtm_virtual_server.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+            if virtual_server_name and virtual_server_server:
+                if virtual_server_exists(api, virtual_server_name, virtual_server_server):
+                    if not module.check_mode:
+                        remove_virtual_server(api, virtual_server_name, virtual_server_server)
+                        result = {'changed': True}
+                    else:
+                        # check-mode return value
+                        result = {'changed': True}
+        elif state == 'present':
+            if virtual_server_name and virtual_server_server and address and port:
```
Comment (srgvg) [inline network/f5/bigip_gtm_virtual_server.py:197]: What happens when address and/or port are not set?

**Comment by srgvg** [comment]
So, same for me, I don't have GTM gear to test this.
I did review the code to the best of my knowlegde but without begin able to test run it...

Most things seem OK, some minor remarks on the specific functions, but unless I read main() badly (it's late here) it seems to have some logic flaws.

File: network/f5/bigip_gtm_virtual_server.py (hunk 1 above)
Comment (perzizzle) [inline network/f5/bigip_gtm_virtual_server.py]: I think I just copied this out of the LTM modules but seems like a reasonable change to improve readability.

**Comment by perzizzle** [comment]
Thanks for the review, I'll make some updates based on the comments. You can tell I started out with just providing the ability to enable/disable virtual servers and then shoehorned in the ability to add/remove virtual servers.

**Comment by perzizzle** [comment]
Playbook to validate functionality https://gist.github.com/perzizzle/5ffa93735881ce7cbf9b depends on an extra_vars.yml similar to https://gist.github.com/perzizzle/190195da598a0311ce6b run with ansible-playbook main_virtual_server.yml --extra-vars=@extra_vars.yml -vv

**Comment by gregdek** [comment]
@caphrim007 Please review this PR to make sure it adheres to the following guidelines:

http://docs.ansible.com/developing_modules.html#module-checklist

If it passes these guidelines, and if you believe it’s a good PR otherwise, please add a comment with "shipit" in the text, and we will flag it for inclusion.

If it fails these guidelines, please add a comment with "needs_revision" in the text, along with the needed changes, and we will flag it for revision.

File: network/f5/bigip_gtm_virtual_server.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+                    result = {'changed': True}
+
+    except Exception, e:
+        module.fail_json(msg="received exception: %s" % e)
+
+    module.exit_json(**result)
+
+# import module snippets
+from ansible.module_utils.basic import *
+main()
```
Comment (caphrim007) [inline network/f5/bigip_gtm_virtual_server.py]: My only nit here is that I'd wrap this main in a conditional

**Comment by caphrim007** [comment]
I've verified that this module works on 11.6.0 provided you create a data center and a server to associate to the VS. Add the conditional to the bottom and lgtm

**Comment by robynbergeron** [comment]
Hi @perzizzle -- sorry this one is taking so long -- looks like it fell through the cracks. :\

Looks like @caphrim007 had a request for you to add something -- can you address that, and then it looks like we can probably merge this? Thanks!

**Comment by caphrim007** [comment]
@robynbergeron  arg, my bad. I forgot to tag the needs_revision. Still new to this workflow. sorry about that!

**Comment by gregdek** [comment]
@caphrim007 no worries; so are we. :)

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

**Comment by perzizzle** [comment]
@gregdek @caphrim007 I updated the module to check if the server exists before attempting to create a virtual server. I think this is all you were looking for...

**Comment by caphrim007** [comment]
I will re-test shortly. thanks @perzizzle

**Comment by caphrim007** [comment]
@gregdek @perzizzle tested and verified it creates, deletes, enables and disables GTM virtual servers. shipit

**Comment by perzizzle** [comment]
@gregdek anything else need to be done here?

**Comment by gregdek** [comment]
@perzizzle Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If we don't hear from you within another 14 days, we will close this pull request.

**Comment by perzizzle** [comment]
@gregdek I don't think there is any needed revision, see previous comment from caphrim. Commit c1fec3f addressed his concern.  He tested and confirmed it behaves as expected and gave it shipit. I don't think I have the permissions necessary to remove the needs_revision label.

**Comment by caphrim007** [comment]
@gregdek shipit ship_it it_puts_the_module_in_the_ship

**Comment by srgvg** [comment]
Please, no container jokes.

File: network/f5/bigip_gtm_virtual_server.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+        required: true
+    password:
+        description:
+            - BIG-IP password
+        required: true
+    state:
+        description:
+            - Virtual server state
+        required: true
+        choices: ['present', 'absent','enabled','disabled']
```
Comment (resmo) [inline network/f5/bigip_gtm_virtual_server.py:55]: many modules uses present as default and it became a convention. would you mind using present as default and set required to false?

File: network/f5/bigip_gtm_virtual_server.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+DOCUMENTATION = '''
+---
+module: bigip_gtm_virtual_server
+short_description: "Manages F5 BIG-IP GTM virtual servers"
+description:
+    - "Manages F5 BIG-IP GTM virtual servers"
+version_added: "2.0"
```
Comment (resmo) [inline network/f5/bigip_gtm_virtual_server.py]: please update to "2.1"

File: network/f5/bigip_gtm_virtual_server.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+            - Virtual server name
+        required: True
+    virtual_server_server:
+        description:
+            - Virtual server server
+        required: true
+    host:
+        description:
+            - Virtual server host
+        required: false
```
Comment (resmo) [inline network/f5/bigip_gtm_virtual_server.py]: if you an arg to required: false, a default must be specified even the default is null/None

File: network/f5/bigip_gtm_virtual_server.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+    virtual_server_id = {'name': name, 'server': server}
+    state = "STATE_%s" % state.strip().upper()
+    api.GlobalLB.VirtualServerV2.set_enabled_state([virtual_server_id], [state])
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            server = dict(type='str', required=True),
+            user = dict(type='str', required=True),
+            password = dict(type='str', required=True),
```
Comment (resmo) [inline network/f5/bigip_gtm_virtual_server.py]: please add a `no_log=True` to password args.

File: network/f5/bigip_gtm_virtual_server.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+            state = dict(type='str', required=True, choices=['present', 'absent', 'enabled', 'disabled']),
```
Comment (resmo) [inline network/f5/bigip_gtm_virtual_server.py]: as mentioned, remove required=True and add default='present' if possible

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by gregdek** [comment]
@perzizzle Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by perzizzle** [comment]
I'll update the PR per the commentary

**Comment by gregdek** [comment]
@perzizzle A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by perzizzle** [comment]
@gregdek @caphrim007 I updated this PR with the parameter changes suggested by @resmo. And then ran through creating, deleting, enabling and disabling virtual servers with the new defaults.

**Comment by resmo** [comment]
there is a catchall exception handling, but not a reason to block this module for integration. Based on my code review, I give it a +1.

shipit

**Comment by gregdek** [comment]
I count multiple shipits. :)

**Comment by resmo** [comment]
@perzizzle as I see you do not return anything, you can just add an empty return:

```
RETURN = '''# '''
```

**Comment by alikins** [comment]
Just to verify, the 'bigip_gtm_virtual_server' module here is different than the https://github.com/ansible/ansible-modules-extras/blob/devel/network/f5/bigip_virtual_server.py
(http://docs.ansible.com/ansible/bigip_virtual_server_module.html) ?

**Comment by perzizzle** [comment]
@alikins that module manages F5 LTMs, this one manages GTM.

**Comment by alikins** [comment]
Pushed to devel at 677a2dd66d5e1788eacc5d849ce6e1cc7d799833 with some minor pep8 cleanups and the RESULTS change resmo suggested.

**Comment by caphrim007** [comment]
@alikins the thing that you pushed to devel went in to the f5/ directory. Not the network/f5 directory as per the other f5 modules. Can you take a look?

**Comment by mhite** [comment]
Yeah, that's weird. Wrong directory!

**Comment by alikins** [comment]
Huh, weird. I think I must have done a git-format-patch/git-am when rebasing/testing and dropped a '-P' level somewhere. 

Fixing...

**Comment by alikins** [comment]
Dir fixup pr at https://github.com/ansible/ansible-modules-extras/pull/2302

**Pull request closed without merging**